Framework/Nexus/inc/MantidNexus/NexusIOHelper.h

```diff
@@ -0,0 +1,162 @@
+#ifndef NEXUSIOHELPER_H
+#define NEXUSIOHELPER_H
+
+#include "MantidIndexing/DllConfig.h"
+#include <algorithm>
+#include <boost/any.hpp>
+#include <nexus/NeXusFile.hpp>
+#include <numeric>
+#include <string>
+
+namespace Mantid {
+namespace NeXus {
+namespace NeXusIOHelper {
+
+namespace {
+
+/// Macro to run a function depending on the type of data in the Nexus file
+#define RUN_NEXUSIOHELPER_FUNCTION(type, func_name, ...)                       \
+  switch (type) {                                                              \
+  case ::NeXus::FLOAT32:                                                       \
+    return func_name<T, float>(__VA_ARGS__);                                   \
+  case ::NeXus::FLOAT64:                                                       \
+    return func_name<T, double>(__VA_ARGS__);                                  \
+  case ::NeXus::INT8:                                                          \
+    return func_name<T, int8_t>(__VA_ARGS__);                                  \
+  case ::NeXus::UINT8:                                                         \
+    return func_name<T, uint8_t>(__VA_ARGS__);                                 \
+  case ::NeXus::INT16:                                                         \
+    return func_name<T, int16_t>(__VA_ARGS__);                                 \
+  case ::NeXus::UINT16:                                                        \
+    return func_name<T, uint16_t>(__VA_ARGS__);                                \
+  case ::NeXus::INT32:                                                         \
+    return func_name<T, int32_t>(__VA_ARGS__);                                 \
+  case ::NeXus::UINT32:                                                        \
+    return func_name<T, uint32_t>(__VA_ARGS__);                                \
+  case ::NeXus::INT64:                                                         \
+    return func_name<T, int64_t>(__VA_ARGS__);                                 \
+  case ::NeXus::UINT64:                                                        \
+    return func_name<T, uint64_t>(__VA_ARGS__);                                \
+  default:                                                                     \
+    throw std::runtime_error("Unknown type in Nexus file");                    \
+  }
+
+std::pair<::NeXus::Info, bool> checkIfOpenAndGetInfo(::NeXus::File &file,
+                                                     std::string entry) {
+  std::pair<::NeXus::Info, bool> info_and_close;
+  info_and_close.second = false;
+  if (!file.isDataSetOpen()) {
+    file.openData(entry);
+    info_and_close.second = true;
+  }
+  info_and_close.first = file.getInfo();
+  return info_and_close;
+}
+
+/// Use the getData function to read the buffer and close file if needed
+template <typename T>
+void callGetData(::NeXus::File &file, std::vector<T> &buf, bool close_file) {
+  file.getData(buf);
+  if (close_file)
```

Comment:
FYI, the file will remain open if `file.getData(buf)` throws. I'm not sure it's worth the effort to implement/use a smart pointer approach to guarantee closing. Thought i'd point it out anyway.